### ISSUE TITLE
Automatically handle child controller insetting

### ISSUE THREAD

**Issue by msaps** [uias/Tabman#43]
In addition to the `requiredContentInset` property in `TabmanBarConfig`, Tabman should automatically handle insetting child view controllers within the page view controller. 

This should also be able to be disabled via an opt-out property in `TabmanViewController`.

**Comment by msaps** [comment]
Issue first suggest in #48

**Comment by msaps** [comment]
Reopening to centralise child insetting discussion.

**Comment by msaps** [comment]
This is now available as of [0.5.0 Beta 1](https://github.com/uias/Tabman/releases/tag/0.5.0.beta.1). 

It is enabled by default but can be disabled via the `automaticallyInsetsChildScrollViews` property on `TabmanViewController`. Any `UITableView` or `UICollectionView` properties should automatically be insetted below the `TabmanBar`.

**Comment by af-esc** [comment]
The new properties won't fix the issue. I use a NavigationController as root. The first view controller of the NavigationController is the TabmanViewController. When I implement your example code I get the follow issue: the inset of the first view controller of TabmanViewController isn't correct. It's an issue with the timing to set the datasource. When you set it in viewDidAppear then looks fine. For viewDidLoad or viewWillAppear it isn't so. I think it's a problem with the layout life cycle.

**Comment by msaps** [comment]
@af-esc discovered the issue, it's with `automaticallyAdjustsScrollViewInsets`. Setting that to false on your view controller should solve the issue. 

I'll look into getting this automatically handled by Tabman in the next version.

**Comment by af-esc** [comment]
Thank's, it works fine.

**Comment by msaps** [comment]
I've just released [0.5.0 Beta 2](https://github.com/uias/Tabman/releases/tag/0.5.0.beta.2) which fixes the issues associated with `automaticallyAdjustsScrollViewInsets`. There's also a bit of refactoring from the last release, so please check out the release notes.

**Comment by robcecil** [comment]
I have children view controllers that are not TableViews or CollectionViews. What is the proper technique to offset them so their entire content is visible? I tried setting the

`self.bar.requiredContentInset = ... `

But it appears to be being set elsewhere..
Thanks

**Comment by eman6576** [comment]
Need help with that as well

**Comment by robcecil** [comment]
Btw, I noticed the code:

internal extension TabmanViewController {
    
```
    /// Automatically inset any table/collection views in a child view controller for the TabmanBar.
    ///
    /// - Parameter childViewController: The child view controller.
    func insetChildViewControllerIfNeeded(_ childViewController: UIViewController?) {
        
        guard let childViewController = childViewController else { return }
        guard self.automaticallyAdjustsChildScrollViewInsets else { return }
        
        // if a scroll view is found in child VC subviews inset by the required content inset.
        for subview in childViewController.view?.subviews ?? [] {
            guard subview is UICollectionView || subview is UITableView else { continue }
```

What about classes that derive from UIScrollView, but do not extend UITableView nor UICollectionView?

**Comment by msaps** [comment]
@robcecil @eman6576 The `requiredContentInset` property provides the inset value required to inset something underneath the bar (basically just gives the height value of the visible bar). This doesn't account for any UIKit components, such as `UINavigationBar` or `UITabBar`.

Regarding the automatic insetting, I'm still working on it and will be adding support for all `UIScrollView` components rather than just `UICollectionView` / `UITableView`. I'm also aiming to improve the properties that are available for manual insetting.

**Comment by msaps** [comment]
@robcecil @eman6576 done a bit more work this evening on the insetting behaviour, and I've just released [0.5.0 Beta 3](https://github.com/uias/Tabman/releases/tag/0.5.0.beta.3). Finally starting to be happy with the release, so stable shouldn't be too far away.

Beta 3 contains numerous improvements:
- All scroll views that are contained within a child view controller are now eligible for automatic insetting.
     - Providing that they are pinned to the top or bottom of the root view.
- Added a new `requiredInsets` `TabmanBarInsets` object.
     - A useful struct that provides individual and compounded inset information.
     - Contains the following properties:
          - `.topLayoutGude` (`CGFloat`) - The length of the top layout guide.
          - `.bottomLayoutGude` (`CGFloat`) - The length of the bottom layout guide.
          - `.bar` (`CGFloat`) - The height of the bar.
          - `.all` (`UIEdgeInsets)` - Compounded insets that can be used directly on `UIScrollView` `contentInset`.

Please do check it out 😄

**Comment by msaps** [comment]
Released [0.5.0](https://github.com/uias/Tabman/releases/tag/0.5.0). 🍾 

Nothing has changed from Beta 3, other than I've deprecated `requiredContentInset`.
